client/allocrunner/task_hook_coordinator.go

```diff
@@ -0,0 +1,103 @@
+package allocrunner
+
+import (
+	"context"
+	"fmt"
+
+	hclog "github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/nomad/nomad/structs"
+)
+
+// TaskHookCoordinator helps coordinates when main start tasks can launch
+// namely after all Prestart Tasks have run, and after all BlockUntilCompleted have completed
+type taskHookCoordinator struct {
+	logger hclog.Logger
+
+	closedCh chan struct{}
+
+	mainTaskCtx       context.Context
+	mainTaskCtxCancel func()
+
+	prestartTasksUntilRunning   map[string]struct{}
+	prestartTasksUntilCompleted map[string]struct{}
+}
+
+func newTaskHookCoordinator(logger hclog.Logger, tasks []*structs.Task) *taskHookCoordinator {
+	closedCh := make(chan struct{})
+	close(closedCh)
+
+	mainTaskCtx, cancelFn := context.WithCancel(context.Background())
+
+	c := &taskHookCoordinator{
+		logger:                      logger,
+		closedCh:                    closedCh,
+		mainTaskCtx:                 mainTaskCtx,
+		mainTaskCtxCancel:           cancelFn,
+		prestartTasksUntilRunning:   map[string]struct{}{},
+		prestartTasksUntilCompleted: map[string]struct{}{},
+	}
+	c.setTasks(tasks)
+	return c
+}
+
+func (c *taskHookCoordinator) setTasks(tasks []*structs.Task) {
+	for _, task := range tasks {
+		if task.Lifecycle == nil || task.Lifecycle.Hook != structs.TaskLifecycleHookPrestart {
+			// move nothing
+			continue
+		}
+
+		// only working with prestart hooks here
+		switch task.Lifecycle.BlockUntil {
+		case "", structs.TaskLifecycleBlockUntilRunning:
```

Comment:
Could we maybe handle the empty string when Canonicalizing the job? - This would mean that we an populate the default in one place, and existing jobs won't be affected by changing defaults etc

Reply:
+1 - this is already being checked in Canonlization and Validation, so don't need to check it here.